test/index.ts

```diff
@@ -2150,4 +2160,125 @@ describe('Datastore', () => {
       assert.strictEqual(key.name, 'Test');
     });
   });
+
+  describe('Without using mocks', () => {
+    const onSaveTests = [
+      {
+        // When the property is contained in excludeFromIndexes
+        // then save should encode request without excludeFromIndexes
+        properties: {k: {stringValue: 'v'}},
+        entitiesWithoutKey: {data: {k: 'v'}},
+      },
+      {
+        // When the property is contained in excludeFromIndexes
+        // then save should ignore non-existent property in excludeFromIndexes
+        properties: {k: {stringValue: 'v', excludeFromIndexes: true}},
+        entitiesWithoutKey: {
+          data: {k: 'v'},
+          excludeFromIndexes: ['k', 'k.*'],
+        },
+      },
+      {
+        // When the property is not contained in excludeFromIndexes
+        // then save should encode a request without excludeFromIndexes
+        properties: {},
+        entitiesWithoutKey: {data: {}},
+      },
+      {
+        // When the property is not contained in excludeFromIndexes
+        // then save should ignore non-existent property in excludeFromIndexes
+        properties: {},
+        entitiesWithoutKey: {
+          data: {},
+          excludeFromIndexes: [
+            'non_exist_property', // this just ignored
+            'non_exist_property.*', // should also be ignored
+          ],
+        },
+      },
+      {
+        // When the property is not contained in excludeFromIndexes and the property is an array
+        // then save should encode a request properly when there is no wildcard
+        properties: {k: {stringValue: 'v'}},
+        entitiesWithoutKey: {
+          data: {k: 'v'},
+          excludeFromIndexes: [
+            'non_exist_property[]', // this just ignored
+          ],
+        },
+      },
+      {
+        // When the property is not contained in excludeFromIndexes and the property is an array
+        // then save should encode a request properly using a wildcard
+        properties: {k: {stringValue: 'v'}},
+        entitiesWithoutKey: {
+          data: {k: 'v'},
+          excludeFromIndexes: [
+            'non_exist_property[].*', // this just ignored
+          ],
+        },
+      },
+    ];
+
+    async.each(
+      onSaveTests,
+      (onSaveTest: {
+        properties: google.datastore.v1.IValue;
+        entitiesWithoutKey: Entities;
+      }) => {
+        it(`should pass the right properties to upsert on save with parameters: ${JSON.stringify(
+          onSaveTest
+        )}`, async () => {
```

Comment:
What if we added a `description` with a human-readable, clear description of each scenario (it's probably fine to just reuse the wording for the comments on top of each object definition) instead of just dropping the blob of `JSON.stringify()` which I'm afraid will get really unreadable and harder to maintain with time.

Reply:
Sounds good. Then we can use the description as the test name.

Reply:
Done